### PR TITLE
Add "32x1 transposed" variant to MXFP8 3D quantization kernel

### PR DESCRIPTION
**Pull request by alexsamardzic**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* __->__ #4356

### CODE REVIEW

**Comment by alexsamardzic** [comment]
Benchmarking results:

```
input_shape       scaling_mode                variant      cuda_2d_us    cutedsl_3d_us    to_mx_us    cuda_2d_gbps    cutedsl_3d_gbps    to_mx_gbps
----------------  --------------------------  ---------  ------------  ---------------  ----------  --------------  -----------------  ------------
(1, 8192, 5120)   ScaleCalculationMode.FLOOR  32x1_t          nan               34.816      26.624         nan                3651.76      4775.39
(1, 8192, 5120)   ScaleCalculationMode.FLOOR  32x1_n           30.72            28.736     118.784        4138.67             4424.41      1070.35
(1, 8192, 5120)   ScaleCalculationMode.FLOOR  32x32_n         nan               30.72       73.696         nan                4138.67      1725.19
(1, 8192, 5120)   ScaleCalculationMode.RCEIL  32x1_t          nan               34.816      36.864         nan                3651.76      3448.89
(1, 8192, 5120)   ScaleCalculationMode.RCEIL  32x1_n           28.672           28.672     120.864        4434.29             4434.29      1051.92
(1, 8192, 5120)   ScaleCalculationMode.RCEIL  32x32_n         nan               28.704      73.728         nan                4429.34      1724.44
(1, 7168, 2048)   ScaleCalculationMode.FLOOR  32x1_t          nan               16.384      59.392         nan                2716          749.241
(1, 7168, 2048)   ScaleCalculationMode.FLOOR  32x1_n           16.384           14.336      83.968        2716                3104          529.951
(1, 7168, 2048)   ScaleCalculationMode.FLOOR  32x32_n         nan               14.336      69.632         nan                3104          639.059
(1, 7168, 2048)   ScaleCalculationMode.RCEIL  32x1_t          nan               16.384      63.488         nan                2716          700.903
(1, 7168, 2048)   ScaleCalculationMode.RCEIL  32x1_n           14.368           14.336      86.016        3097.09             3104          517.333
(1, 7168, 2048)   ScaleCalculationMode.RCEIL  32x32_n         nan               14.368      71.648         nan                3097.09       621.077
(8, 8192, 5120)   ScaleCalculationMode.FLOOR  32x1_t          nan              186.368    1208.48          nan                5457.58       841.651
(8, 8192, 5120)   ScaleCalculationMode.FLOOR  32x1_n         1524.74           169.984    1689.63          667.079            5983.61       601.976
(8, 8192, 5120)   ScaleCalculationMode.FLOOR  32x32_n         nan              174.08     1816.58          nan                5842.82       559.91
(8, 8192, 5120)   ScaleCalculationMode.RCEIL  32x1_t          nan              184.288    1289.22          nan                5519.18       788.944
(8, 8192, 5120)   ScaleCalculationMode.RCEIL  32x1_n         1506.27           163.872    1764.35          675.256            6206.79       576.483
(8, 8192, 5120)   ScaleCalculationMode.RCEIL  32x32_n         nan              166.976    1672.19          nan                6091.41       608.255
(8, 7168, 2048)   ScaleCalculationMode.FLOOR  32x1_t          nan               70.656     433.2           nan                5038.38       821.772
(8, 7168, 2048)   ScaleCalculationMode.FLOOR  32x1_n          540.704           65.536     602.112         658.385            5432          591.238
(8, 7168, 2048)   ScaleCalculationMode.FLOOR  32x32_n         nan               67.584     643.072         nan                5267.39       553.58
(8, 7168, 2048)   ScaleCalculationMode.RCEIL  32x1_t          nan               69.664     458.752         nan                5110.12       776
(8, 7168, 2048)   ScaleCalculationMode.RCEIL  32x1_n          534.56            65.536     626.72          665.952            5432          568.023
(8, 7168, 2048)   ScaleCalculationMode.RCEIL  32x32_n         nan               65.536     593.92          nan                5432          599.393
(32, 7168, 2048)  ScaleCalculationMode.FLOOR  32x1_t          nan              252        1686.56          nan                5650.66       844.302
(32, 7168, 2048)  ScaleCalculationMode.FLOOR  32x1_n         2128.9            235.52     2368.54          668.875            6046.05       601.199
(32, 7168, 2048)  ScaleCalculationMode.FLOOR  32x32_n         nan              245.792    2544.61          nan                5793.38       559.601
(32, 7168, 2048)  ScaleCalculationMode.RCEIL  32x1_t          nan              247.84     1807.36          nan                5745.51       787.871
(32, 7168, 2048)  ScaleCalculationMode.RCEIL  32x1_n         2102.24           223.392    2471.97          677.357            6374.29       576.046
(32, 7168, 2048)  ScaleCalculationMode.RCEIL  32x32_n         nan              231.456    2354.19          nan                6152.21       604.864
(32, 8192, 5120)  ScaleCalculationMode.FLOOR  32x1_t          nan              740.912    4806.78          nan                5491.17       846.403
(32, 8192, 5120)  ScaleCalculationMode.FLOOR  32x1_n         6059.01           666.624    6730.75          671.475            6103.1        604.461
(32, 8192, 5120)  ScaleCalculationMode.FLOOR  32x32_n         nan              708.576    7233.54          nan                5741.76       562.446
(32, 8192, 5120)  ScaleCalculationMode.RCEIL  32x1_t          nan              718.848    5125.71          nan                5659.72       793.738
(32, 8192, 5120)  ScaleCalculationMode.RCEIL  32x1_n         5980.18           633.792    7013.41          680.327            6419.26       580.1
(32, 8192, 5120)  ScaleCalculationMode.RCEIL  32x32_n         nan              649.216    6655.12          nan                6266.75       611.33
```

**Comment by danielvegamyhre** [comment]
@alexsamardzic can you benchmark this against the 2 stage approach we do here in `_compute_fwd_sm100()`: https://github.com/pytorch/ao/blob/9052ece233a1939747d5c1ee46edca40f53aa152/torchao/prototype/moe_training/mxfp8_grouped_mm.py#L555-L558

**Comment by alexsamardzic** [comment]
> can you benchmark this against the 2 stage approach we do here in `_compute_fwd_sm100()`:
> 
> https://github.com/pytorch/ao/blob/9052ece233a1939747d5c1ee46edca40f53aa152/torchao/prototype/moe_training/mxfp8_grouped_mm.py#L555-L558

Here is an adapted benchmarking script to compare between the two: [bench_quantize_3d_vs_triton.py](https://github.com/user-attachments/files/27251261/bench_quantize_3d_vs_triton.py).

And here are the results:

```
input_shape       scaling_mode                  cutedsl_3d_us    triton_two_stage_us  speedup
----------------  --------------------------  ---------------  ---------------------  ---------
(1, 8192, 5120)   ScaleCalculationMode.FLOOR           34.816                 34.848  1.00x
(1, 8192, 5120)   ScaleCalculationMode.RCEIL           32.832                 34.848  1.06x
(1, 7168, 2048)   ScaleCalculationMode.FLOOR           16.224                 20.736  1.28x
(1, 7168, 2048)   ScaleCalculationMode.RCEIL           14.336                 20.352  1.42x
(8, 8192, 5120)   ScaleCalculationMode.FLOOR          186.4                  196.608  1.05x
(8, 8192, 5120)   ScaleCalculationMode.RCEIL          184.384                198.656  1.08x
(8, 7168, 2048)   ScaleCalculationMode.FLOOR           71.68                  72.72   1.01x
(8, 7168, 2048)   ScaleCalculationMode.RCEIL           69.632                 73.728  1.06x
(32, 7168, 2048)  ScaleCalculationMode.FLOOR          261.12                 285.632  1.09x
(32, 7168, 2048)  ScaleCalculationMode.RCEIL          256                    287.712  1.12x
(32, 8192, 5120)  ScaleCalculationMode.FLOOR          759.808               1029.09   1.35x
(32, 8192, 5120)  ScaleCalculationMode.RCEIL          737.312                990.304  1.34x
```

File: benchmarks/prototype/moe_training/mxfp8/roofline_unified.py [hunk 1]
```diff
@@ -441,8 +432,6 @@ def benchmark_mxfp8_grouped_mm_fwd_bwd(x, w_t, offs, labels):
     x_clone = x.clone().requires_grad_(True)
     w_t_clone = w_t.clone().requires_grad_(True)
 
-    fn = torch.compile(_to_mxfp8_then_scaled_grouped_mm, fullgraph=True)
```
Comment (danielvegamyhre) [inline benchmarks/prototype/moe_training/mxfp8/roofline_unified.py:444]: why was compile removed here?

File: test/prototype/moe_training/test_mxfp8_grouped_mm.py [hunk 2]
```diff
@@ -60,58 +60,62 @@
...
-    block_size = 32
-    x_scale, x_fp8 = to_mx(
-        x,
-        elem_dtype=torch.float8_e4m3fn,
+    if variant == "32x1_t":
+        # Forward-style grouped MM: input @ weight.transpose(-2, -1).
+        input_act = torch.randn(M, K, dtype=torch.bfloat16, device="cuda")
+        weight = torch.randn(num_experts, N, K, dtype=torch.bfloat16, device="cuda")
+        mat2 = weight.transpose(-2, -1)
+        scale_block_k = 1
```
Comment (danielvegamyhre) [inline test/prototype/moe_training/test_mxfp8_grouped_mm.py]: as an aside, in the future we should probably refactor to not use the "scale_block_n" / "scale_block_k" naming everywhere since that leads to cases like this where we are quantizing pre-transposed input shape (E,K,N) along K, yet setting scale_block_k=1 for api consistency.

maybe in a follow up we can refactor call them scale_block_dim1, scale_block_dim2 or something?

File: test/prototype/moe_training/test_mxfp8_grouped_mm.py [hunk 3]
```diff
@@ -60,58 +60,62 @@
...
+        scale_block_k=scale_block_k,
+        scaling_mode=scale_mode.value.lower(),
+        blocked_scale_output=False,
     )
-    x_ref = (
-        x_fp8.reshape(M, K // block_size, block_size).to(torch.bfloat16)
-        * x_scale.unsqueeze(-1).to(torch.bfloat16)
-    ).reshape(M, K)
 
+    input_scale_ref = input_scale.repeat_interleave(block_size, dim=1)
```
Comment (danielvegamyhre) [inline test/prototype/moe_training/test_mxfp8_grouped_mm.py:100]: why is the LHS input activation (scaled with 1x32) being repeat interleaved here? i would think repeat interleave would only be necessary for replicating the scale for a 32x32 weight scaling reference impl?

File: torchao/prototype/moe_training/kernels/mxfp8/cutedsl_quantize_3d.py [hunk 4]
```diff
@@ -471,33 +488,41 @@ def kernel(
...
-                stride=(TILE_N * TILE_K, TILE_N * TILE_K, TILE_K, 1),
+                cute,
+                TILE_N,
+                TILE_K,
+                INPUT_TRANSPOSED_VALUE,
             )
+            if cutlass.const_expr(INPUT_TRANSPOSED_VALUE):
+                staged_layout_in = cute.make_layout(
+                    (STAGE_COUNT_VALUE, 1, TILE_N, TILE_K),
+                    stride=(STAGE_ELEMS, STAGE_ELEMS, 1, TILE_N),
```
Comment (danielvegamyhre) [inline torchao/prototype/moe_training/kernels/mxfp8/cutedsl_quantize_3d.py]: cutedsl question: why is "num stages" instances of TILE_NxTILE_K tiles represented as a 4d tensor of shape (stages, 1, tile_n, tile_k), rather than a 3d tensor of shape (stages, tile_n, tile_k)?

File: torchao/prototype/moe_training/kernels/mxfp8/cutedsl_quantize_3d.py [hunk 5]
```diff
@@ -21,11 +21,22 @@
 )
 
 
-def _make_tile_smem_layouts(cute, tile_n: int, tile_k: int):
-    smem_layout_in = cute.make_layout(
-        (1, tile_n, tile_k),
-        stride=(tile_n * tile_k, tile_k, 1),
-    )
+def _make_tile_smem_layouts(
+    cute,
```
Comment (danielvegamyhre) [inline torchao/prototype/moe_training/kernels/mxfp8/cutedsl_quantize_3d.py]: why does cute package need to be a param here, is it due to doing the import with guards elsewhere instead of at the top? not a huge deal but this feels a bit awkward

**Review by danielvegamyhre — APPROVED** [review]
LGTM with some minor comments/questions

File: benchmarks/prototype/moe_training/mxfp8/roofline_unified.py (hunk 1 above)
Comment (alexsamardzic) [reply benchmarks/prototype/moe_training/mxfp8/roofline_unified.py:444]: Wrong edit, reverted.

File: test/prototype/moe_training/test_mxfp8_grouped_mm.py (hunk 2 above)
Comment (alexsamardzic) [reply test/prototype/moe_training/test_mxfp8_grouped_mm.py]: Indeed these are better names,  I did the rename now.

File: test/prototype/moe_training/test_mxfp8_grouped_mm.py (hunk 3 above)
Comment (alexsamardzic) [reply test/prototype/moe_training/test_mxfp8_grouped_mm.py:100]: This should be just dequantization for the BF16 reference, scales are (M, dim//32) so we expand to (M, dim).

File: torchao/prototype/moe_training/kernels/mxfp8/cutedsl_quantize_3d.py (hunk 4 above)
Comment (alexsamardzic) [reply torchao/prototype/moe_training/kernels/mxfp8/cutedsl_quantize_3d.py]: I mechanically used (stages, 1, tile_n, tile_k) to mirror the per-stage TMA tile shape (1, tile_n, tile_k). The singleton dim indeed is not needed at all, so it's removed now.

File: torchao/prototype/moe_training/kernels/mxfp8/cutedsl_quantize_3d.py (hunk 5 above)
Comment (alexsamardzic) [reply torchao/prototype/moe_training/kernels/mxfp8/cutedsl_quantize_3d.py]: This is a remnant of first try to make code working in case when CuTeDSL package is not installed, replaced with a simple import within the body (also made the change for alike methods for 2D kernels).

**Comment by alexsamardzic** [comment]
To be merged into main as #4383.
